### DISCUSSION TITLE
How to build a nice interactive Textual representation of a hierarchical parameter container ?

### DISCUSSION THREAD

**Discussion by paugier** [Textualize/textual#4318] [Q&A]
I like to use Textual to build small apps to launch/control/post-process computations using [Fluiddyn](https://fluiddyn.readthedocs.io) Python libraries. The API is often based on instances of Fluiddyn [ParamContainer](https://fluiddyn.readthedocs.io/en/latest/generated/fluiddyn.util.paramcontainer.html)

For example, [this tutorial](https://fluidimage.readthedocs.io/en/latest/tutorials/tuto_piv.html) shows how to compute Particle Image Velocimetry with Fluidimage:

```python
from fluidimage.piv import Work

params = Work.create_default_params()
# the user can interact with the object `params` to
# - read the default parameters
# - read the documentation on parameters
# - change the parameters

params.series.path = ...
params.piv0.shape_crop_im0 = 64
params.multipass.number = 2

# params is a hierarchical object, a `ParamContainer` containing other `ParamContainer` objects (`series`, `piv0`, `multipass`, etc.)

work = Work(params)
result = work.process_1_serie()
```

I'd like to be able to produce a nice Textual representation of a `ParamContainer` to be able to graphically interact with it (discover the possible parameters, read the documentation and of course modify the parameters).

I see that there are plenty of nice widgets that could be used for different types of parameters. I could even add data in the `ParamContainer` object on which widget should be used for which parameter.

However, right now I don't see how to handle in Textual the hierarchical nature of  `ParamContainer` objects. So here is my question:

How could one represent in Textual hierarchical Python objects containing parameters ?

**Comment by davep** [comment]
While looking at all of the widgets, did you consider [`Tree`](https://textual.textualize.io/widgets/tree/)?

**Reply by paugier** [reply]
If I understand correctly, `Tree`s are designed "for displaying and navigating data in a tree", but I can't add widgets in a tree to be able to modify the data. So it seems that I would need a bit more than a `Tree`.

**Reply by davep** [reply]
Correct, you can't place widgets within a `Tree`, but you could use a `Tree` to navigate the properties and have an adjacent container of widgets that pertain to the property that is selected (as one approach).
